java/src/main/native/ai_onnxruntime_OrtSession.c

```diff
@@ -242,116 +301,164 @@ JNIEXPORT jobjectArray JNICALL Java_ai_onnxruntime_OrtSession_getOutputInfo
-    for (int i = 0; i < numOutputs; i++) {
-        javaOutputStrings[i] = (*jniEnv)->GetObjectArrayElement(jniEnv,outputNamesArr,i);
-        outputNames[i] = (*jniEnv)->GetStringUTFChars(jniEnv,javaOutputStrings[i],NULL);
-        outputValues[i] = NULL;
-    }
+  // Extract a C array of longs which are pointers to the input tensors.
+  // Need to convert longs to OrtValue* in case we run on non-64bit systems
+  jlong* inputTensors = (*jniEnv)->GetLongArrayElements(jniEnv, tensorArr, NULL);
 
-    // Actually score the inputs.
-    //printf("inputTensors = %p, first tensor = %p, numInputs = %ld, outputValues = %p, numOutputs = %ld\n",inputTensors,(OrtValue*)inputTensors[0],numInputs,outputValues,numOutputs);
-    //ORT_API_STATUS(OrtRun, _Inout_ OrtSession* sess, _In_ OrtRunOptions* run_options, _In_ const char* const* input_names, _In_ const OrtValue* const* input, size_t input_len, _In_ const char* const* output_names, size_t output_names_len, _Out_ OrtValue** output);
-    checkOrtStatus(jniEnv,api,api->Run(session, runOptions, (const char* const*) inputNames, (const OrtValue* const*) inputValues, numInputs, (const char* const*) outputNames, numOutputs, outputValues));
-    // Release the C array of pointers to the tensors.
-    (*jniEnv)->ReleaseLongArrayElements(jniEnv,tensorArr,inputTensors,JNI_ABORT);
-
-    // Construct the output array of ONNXValues
-    char *onnxValueClassName = "ai/onnxruntime/OnnxValue";
-    jclass onnxValueClass = (*jniEnv)->FindClass(jniEnv, onnxValueClassName);
-    jobjectArray outputArray = (*jniEnv)->NewObjectArray(jniEnv,safecast_int64_to_jsize(numOutputs), onnxValueClass, NULL);
-
-    // Convert the output tensors into ONNXValues and release the output strings.
-    for (int i = 0; i < numOutputs; i++) {
-        if (outputValues[i] != NULL) {
-            jobject onnxValue = convertOrtValueToONNXValue(jniEnv,api,allocator,outputValues[i]);
-            (*jniEnv)->SetObjectArrayElement(jniEnv,outputArray,i,onnxValue);
-        }
-        (*jniEnv)->ReleaseStringUTFChars(jniEnv,javaOutputStrings[i],outputNames[i]);
-    }
-    checkOrtStatus(jniEnv,api,api->AllocatorFree(allocator,outputValues));
+  // Extract the names and native pointers of the input values.
+  for (int i = 0; i < numInputs; i++) {
+    javaInputStrings[i] = (*jniEnv)->GetObjectArrayElement(jniEnv, inputNamesArr, i);
+    inputNames[i] = (*jniEnv)->GetStringUTFChars(jniEnv, javaInputStrings[i], NULL);
+    inputValues[i] = (OrtValue*)inputTensors[i];
+  }
+
+  // Release the java array copy of pointers to the tensors.
+  (*jniEnv)->ReleaseLongArrayElements(jniEnv, tensorArr, inputTensors, JNI_ABORT);
+
+  // Extract the names of the output values.
+  for (int i = 0; i < numOutputs; i++) {
+    javaOutputStrings[i] = (*jniEnv)->GetObjectArrayElement(jniEnv, outputNamesArr, i);
+    outputNames[i] = (*jniEnv)->GetStringUTFChars(jniEnv, javaOutputStrings[i], NULL);
+    outputValues[i] = NULL;
```

Comment:
>  outputValues[i] = NULL; [](http://example.com/codeflow?start=2&length=25)

We generally allow user to pass their output OrtValues. This is useful when they want output to go to some pre-allocated memory of their choice.

Reply:
I agree, but that functionality was enabled in the C API long after this code was written, and I need to rewrite all the Java side bits to support it. Let's leave it for another PR. It needs more work to allow long lived byte buffers to ensure the zero copy path out and it'll take some Java side refactoring to enable that.